private/bufpkg/bufimage/bufimageutil/bufimageutil.go

```diff
@@ -381,139 +421,175 @@ func descriptorTransitiveClosure(namedDescriptor protosource.NamedDescriptor, im
 				return nil, fmt.Errorf("unknown field type %d", field.Type())
 			}
 			// fieldoptions
-			explicitOptionDeps, recursedOptionDeps, err := exploreCustomOptions(field, imageIndex, seen)
+			explicitOptionDeps, recursedOptionDeps, err := exploreCustomOptions(field, imageIndex, seen, opts)
 			if err != nil {
 				return nil, err
 			}
 			directDependencies = append(directDependencies, explicitOptionDeps...)
 			transitiveDependencies = append(transitiveDependencies, recursedOptionDeps...)
 		}
-		// Extensions declared for this message
-		for _, extendsDescriptor := range imageIndex.NameToExtensions[namedDescriptor.FullName()] {
-			directDependencies = append(directDependencies, extendsDescriptor)
-			recursiveDescriptors, err := descriptorTransitiveClosure(extendsDescriptor, imageIndex, seen)
-			if err != nil {
-				return nil, err
+		// Extensions declared for this message.
+		// TODO: We currently exclude all extensions for descriptor options types (and instead
+		//  only gather the ones used in relevant custom options). But if the descriptor option
+		//  type were a named type for filtering, we SHOULD include all of them.
```

Comment:
I kind of think of them differently: I think of `ExcludeCustomOptions` as meaning "don't look at custom options on the elements in the image", not "don't ever return extensions that are custom options".

Perhaps I should rename `WithExcludeCustomOptions` to `WithNotExploringCustomOptions`. WDYT?

Reply:
Now that I think about it, it's not just if the descriptor options type is explicitly named as a type filter that we'd want to include their extensions. It's also if the descriptor options type is actually referenced in the graph _other_ than implicitly via custom options. For example, a request message contains `FileDescriptorSet`. Suddenly, all of the descriptor options types are part of the message's schema, even if none of the elements declare custom options.

Reply:
I think it's a weird case that depends on the use. 

If I'm querying for the new `/GetSchema` endpoint where FileDescriptorSet is a dependency, you probably want an option to avoid including custom options declared in scope where `/GetSchema` is declared (buf's custom options). After all, we may declare a methodoption on tag no 50000-99999 (the organization [reserved range](https://developers.google.com/protocol-buffers/docs/proto#customoptions)) that is of type string, but that shouldn't prevent anyone from uploading their schema to us if they've decided that that tag is something else.

Mabye `ExcludeCustomOptionValues` would be more representative for returning a filedescriptor in which concrete custom options aren't returned? Although in that case if there are no values I'd imagine you want to scrap including their definitions as well.

Reply:
We should _not_ be using the organization reserved range for anything that could be externally visible. Otherwise, we clearly run the risk of a conflict.

We should instead reserve tag numbers in the global reserved list. I actually started this a few weeks ago:
https://github.com/protocolbuffers/protobuf/compare/main...jhump:protobuf:jh/custom-options-for-buf-and-connect?expand=1
But I noticed that some reservations grab a range of _ten_ tags. So I was thinking of grabbing ten consecutive numbers for both and for connect, just in case.